From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] setupapi: the local computer name is a machine name (3745)

setupapi: the local computer name is a machine name

SetupDiCreateDeviceInfoListEx, SetupDiGetClassDevsEx and SetupDiOpenClassRegKeyEx
accept NULL, an empty string and the name of this computer (\\name) as the
local machine; other machines are ERROR_INVALID_MACHINENAME in all three
(OpenClassRegKeyEx and a handed-in device set were not checked).
---
diff --git a/dlls/setupapi/devinst.c b/dlls/setupapi/devinst.c
index 3d5a002..f6f612d 100644
--- a/dlls/setupapi/devinst.c
+++ b/dlls/setupapi/devinst.c
@@ -1619,6 +1619,24 @@ SetupDiCreateDeviceInfoList(const GUID *ClassGuid,
   return SetupDiCreateDeviceInfoListExW(ClassGuid, hwndParent, NULL, NULL);
 }
 
+/* Stained Glass OS: a machine name is "\\\\name" (the form of CM_Connect_Machine); the name of this computer is local
+ * (NULL and "" are too).  Any other machine would have to be reached over the network, which is not done. */
+static BOOL is_local_machine_name(const WCHAR *name)
+{
+    static const COMPUTER_NAME_FORMAT formats[] = { ComputerNameNetBIOS, ComputerNameDnsHostname, ComputerNameDnsFullyQualified };
+    WCHAR local[256];
+    DWORD size, i;
+
+    if (!name || !*name) return TRUE;
+    if (name[0] != '\\' || name[1] != '\\' || !name[2]) return FALSE;
+    for (i = 0; i < ARRAY_SIZE(formats); i++)
+    {
+        size = ARRAY_SIZE(local);
+        if (GetComputerNameExW(formats[i], local, &size) && !lstrcmpiW(name + 2, local)) return TRUE;
+    }
+    return FALSE;
+}
+
 /***********************************************************************
  *		SetupDiCreateDeviceInfoListExA (SETUPAPI.@)
  */
@@ -1677,9 +1695,9 @@ SetupDiCreateDeviceInfoListExW(const GUID *ClassGuid,
     TRACE("%s %p %s %p\n", debugstr_guid(ClassGuid), hwndParent,
       debugstr_w(MachineName), Reserved);
 
-    if (MachineName && *MachineName)
+    if (!is_local_machine_name(MachineName))
     {
-        FIXME("remote support is not implemented\n");
+        FIXME("%s: remote support is not implemented\n", debugstr_w(MachineName));
         SetLastError(ERROR_INVALID_MACHINENAME);
         return INVALID_HANDLE_VALUE;
     }
@@ -2859,6 +2877,12 @@ HDEVINFO WINAPI SetupDiGetClassDevsExW(const GUID *class, PCWSTR enumstr, HWND p
         SetLastError(ERROR_INVALID_PARAMETER);
         return INVALID_HANDLE_VALUE;
     }
+    if (!is_local_machine_name(machine))
+    {
+        FIXME("%s: remote support is not implemented\n", debugstr_w(machine));
+        SetLastError(ERROR_INVALID_MACHINENAME);
+        return INVALID_HANDLE_VALUE;
+    }
     if (flags & DIGCF_ALLCLASSES)
         class = NULL;
 
@@ -2888,10 +2912,7 @@ HDEVINFO WINAPI SetupDiGetClassDevsExW(const GUID *class, PCWSTR enumstr, HWND p
         set = SetupDiCreateDeviceInfoListExW((flags & DIGCF_DEVICEINTERFACE) ? NULL : class, parent, machine, reserved);
     if (set != INVALID_HANDLE_VALUE)
     {
-        if (machine && *machine)
-            FIXME("%s: unimplemented for remote machines\n",
-                    debugstr_w(machine));
-        else if (flags & DIGCF_DEVICEINTERFACE)
+        if (flags & DIGCF_DEVICEINTERFACE)
             SETUPDI_EnumerateInterfaces(set, class, enumstr, flags);
         else
             SETUPDI_EnumerateDevices(set, class, enumstr, flags);
@@ -3769,9 +3790,10 @@ HKEY WINAPI SetupDiOpenClassRegKeyExW(
     LPCWSTR lpKeyName;
     LONG l;
 
-    if (MachineName && *MachineName)
+    if (!is_local_machine_name(MachineName))
     {
-        FIXME("Remote access not supported yet!\n");
+        FIXME("%s: remote access not supported yet!\n", debugstr_w(MachineName));
+        SetLastError(ERROR_INVALID_MACHINENAME);
         return INVALID_HANDLE_VALUE;
     }
 
